opentelemetry-api/src/opentelemetry/metrics/instrument.py

```diff
@@ -0,0 +1,204 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=too-many-ancestors
+
+# type: ignore
+
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+from re import compile as compile_
+from threading import RLock
+from types import GeneratorType
+
+from opentelemetry.metrics.measurement import Measurement
+
+_logger = getLogger(__name__)
+
+
+class Instrument(ABC):
+
+    _name_regex = compile_(r"[a-zA-Z][-.\w]{0,62}")
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def unit(self):
+        return self._unit
+
+    @property
+    def description(self):
+        return self._description
+
+    @abstractmethod
+    def __init__(self, name, *args, unit="", description="", **kwargs):
+        self._lock = RLock()
+
+        if name is None or self._name_regex.fullmatch(name) is None:
+            _logger.error("Invalid instrument name %s", name)
+
+        else:
+            self._name = name
+
+        if unit is None:
+            self._unit = ""
+        elif len(unit) > 63:
+            _logger.error("unit must be 63 characters or shorter")
+
+        elif any(ord(character) > 127 for character in unit):
+            _logger.error("unit must only contain ASCII characters")
+        else:
+            self._unit = unit
+
+        if description is None:
+            description = ""
+
+        self._description = description
+
+
+class Synchronous(Instrument):
+    pass
+
+
+class Asynchronous(Instrument):
+    @abstractmethod
+    def __init__(
+        self, name, callback, *args, unit="", description="", **kwargs
+    ):
+        super().__init__(name, *args, unit=unit, description="", **kwargs)
+        with self._lock:
+
+            if not isinstance(callback, GeneratorType):
+                _logger.error("callback must be a generator")
+
+            else:
+                super().__init__(
+                    name, unit=unit, description=description, *args, **kwargs
+                )
+                self._callback = callback
+
+    def observe(self):
```

Comment:
This shouldn't be present in the API

Reply:
We need some interface in the API for the callbacks because we need to do some checking on the returned value of the callback. I understand that the spec says that no operations are to be provided for asynchronous instruments but we still need something. We can hide the fact that we provide an API by defining a method named `callback`, something like this:

``` python
from abc import ABC, abstractmethod
from logging import getLogger

_logger = getLogger(__name__)


class Measurement:
    def __init__(self, value):
        self.value = value


class Asynchronous(ABC):

    def __init__(self, callback):
        self._callback = callback

    @property
    @abstractmethod
    def callback(self):
        def function():
            result = next(callback)
            if not isinstance(result, Measurement):
                _logger.error("result must be a Measurement")
                return None

            return result
        return function


class ObservableCounter(Asynchronous):
    @property
    def callback(self):
        def function():
            result = super(ObservableCounter, self).callback()

            if result is not None and result.value < 0:
                _logger.error("value must be positive")
                return None

            return result
        return function


def callback():
    state = [Measurement(1), Measurement(2), 3, Measurement(-4)]

    while True:
        yield state.pop()


callback = callback()

observable_counter = ObservableCounter(callback)

result = observable_counter.callback()
if result is not None:
    print(result.value)
result = observable_counter.callback()
if result is not None:
    print(result.value)
result = observable_counter.callback()
if result is not None:
    print(result.value)
result = observable_counter.callback()
if result is not None:
    print(result.value)
```
```
value must be positive
result must be a Measurement
2
1
```

Reply:
So how many times do you check the callback to see if it is valid though. I don't think we can reasonably validate the callback's behavior.

Reply:
> So how many times do you check the callback to see if it is valid though.

Hmm, as many times as the callback is called? :thinking: 

> I don't think we can reasonably validate the callback's behavior.

Sorry, why not? I am a bit confused.

Reply:
This is being resolved [here](https://github.com/ocelotl/opentelemetry-python/pull/5).